base/compiler/typeinfer.jl

```diff
@@ -682,7 +642,7 @@ function find_dominating_assignment(id::Int, idx::Int, sv::InferenceState)
 end
 
 # annotate types of all symbols in AST, preparing for optimization
-function type_annotate!(interp::AbstractInterpreter, sv::InferenceState, run_optimizer::Bool)
+function type_annotate!(interp::AbstractInterpreter, sv::InferenceState)
```

Comment:
I'm a bit uncertain about getting rid of this flag, but with these changes the meaning would have been "generate IR that is valid to pass into the optimizer" not just "perform a few optional optimizations".

We could also:
 - add more checks in slot2ssa to deal with `run_optimizer=false`, or else
 - annotate the dead branches for the optimizer to do this modification itself

Reply:
Meh, I think it's fine. It's possible we should just not run this function at all if the optimizer isn't being run, but I think this is fine as is.

Reply:
Can we move `5: replace this live GotoIfNot with:` to the optimizer pass rather than doing it unconditionally here? Then `type_annotate!` really fixes up inference-only types only, which sounds much cleaner.

Reply:
Done!

Reply:
A downside of this approach just occurred to me, which is that running type-inference and then using our [external entrypoint](https://github.com/JuliaLang/julia/pull/50943/files#diff-e04705952aec52ac6b24bd37c25cd11b8ba7d72b488b645bd5d2c104f79ccaabR143) for the optimizer is going be strictly worse than the result we obtain internally, since we won't be able to prune branches or insert PiNodes where we used to have `TypedSlot`.

I don't think there's a way around that, unless we can add this reachability/vartables information to `CodeInfo`?

Reply:
I think that's fine and I am even not sure if there are serious users of the external optimization entrypoint other than the Base test suite. If it turns out to be a problem, we can allow `OptimizationState(::MethodInstance, ::CodeInfo, ::AbstractInterpreter)` to accept optional reachability/variable information.